### ISSUE TITLE
No gradients for bias and kernels in Dense layer. 

### ISSUE THREAD

**Issue by shivam13juna** [tensorflow/tensorflow#36880]
**System information** - Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): - Yes
OS: Mac OS Catalina 
TensorFlow installed from: pip
TensorFlow version:  2.1 
 Python version: - 3.7

I'm getting this error, 

`**WARNING:tensorflow:Gradients do not exist for variables ['flight_2/layer1/kernel:0', 'flight_2/layer1/bias:0', 'flight_2/layer2/kernel:0', 'flight_2/layer2/bias:0', 'flight_2/layer3/kernel:0', 'flight_2/layer3/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['flight_2/layer1/kernel:0', 'flight_2/layer1/bias:0', 'flight_2/layer2/kernel:0', 'flight_2/layer2/bias:0', 'flight_2/layer3/kernel:0', 'flight_2/layer3/bias:0'] when minimizing the loss.**

Here's my full code, which is enough for re-producing problem
```

import dask.dataframe as dd
import numpy as np
import tensorflow as tf
tf.compat.v1.enable_eager_execution()
tf.keras.backend.set_floatx('float64')
import time
pd.options.display.float_format = '{:,.2f}'.format
from scipy import stats
import matplotlib as mpl
mpl.rcParams['agg.path.chunksize'] = 10000
import os




from tensorflow.keras.layers import LSTM, GRU



from sklearn.preprocessing import LabelEncoder, StandardScaler


%matplotlib inline

tf.__version__

# data = dd.read_csv("s3://search-curated-sec/bom_pat_6e_aberogate_new.csv/*.csv")
data = pd.read_csv('../data/bom_pat_6e_abrogate.csv')

data.dropna(inplace=True)

data['deptime'] = data['deptime'].astype(str).str.pad(4, fillchar='0')
data['depinf'] = data['depdate'].astype(str) + data['deptime'].astype(str)
for i in ['booktime','changeDateTime_1', 'changeDateTime_2',
       'changeDateTime_3', 'changeDateTime_4', 'changeDateTime_5',
       'changeDateTime_6', 'changeDateTime_7', 'changeDateTime_8',
       'changeDateTime_9', 'changeDateTime_10', 'aberogateTime']:
    data[i] = pd.to_datetime(data[i], format = '%Y%m%d%H')

data['depinf'] = pd.to_datetime(data['depinf'], format='%Y%m%d%H%M')

data.drop(['bookdate', 'depdate', 'deptime'], axis=1, inplace=True)

data.head()

data['diff_book_3'] = data.apply(lambda x: (x['changeDateTime_3'] - x['booktime']).total_seconds(), axis=1)

data.to_csv('../data/bom_pat_6e_abrogate.csv', index=False)

data[data['diff_book_3']>=0].shape

data.shape

data['diff_book_3'] = (pd.to_datetime(data['changeDateTime_3']) - pd.to_datetime(data['booktime'])).dt.total_seconds() // 3600

### Features

data.columns

lbl1 = LabelEncoder()
std = StandardScaler()

data['trans_fno'] = lbl1.fit_transform(data['fno'])

features = np.array([data['duration'], data['htg'], data['fare'],  data['trans_fno'], pd.to_datetime(data['booktime']).dt.month, pd.to_datetime(data['booktime']).dt.day, pd.to_datetime(data['depinf']).dt.month, pd.to_datetime(data['depinf']).dt.day]).T

target = data['diff_book_3'].values

#Standardizing all features
features = std.fit_transform(features)

# Starting Model

class Flight(tf.keras.Model):
    def __init__(self, layer1_units = 512, layer2_units = 256, layer3_units = 128):
        super(Flight, self).__init__()
        self.layer1 = layer1_units
        self.layer2 = layer2_units
        self.layer3 = layer3_units

        
        self.linear1 = tf.keras.layers.Dense(self.layer1, activation='relu', input_shape = (9, ), kernel_regularizer=tf.keras.regularizers.l2(1e-3), name='layer1')
        self.linear2 = tf.keras.layers.Dense(self.layer2, activation='relu', kernel_regularizer=tf.keras.regularizers.l2(1e-3), name='layer2')
        self.linear3 = tf.keras.layers.Dense(self.layer3, activation='relu', kernel_regularizer=tf.keras.regularizers.l2(1e-3), name='layer3')
        self.linear4 = tf.keras.layers.Dense(1, activation='linear', kernel_regularizer=tf.keras.regularizers.l2(1e-3), name='layer4')


    def call(self, x):
        output = self.linear1(x)
        output = self.linear2(x)
        output = self.linear3(x)
        output = self.linear4(x)
        return output

#### Creating a tf.data dataset 

batch_size = 10000


data_size = len(features)
steps_per_epoch = len(features)//batch_size

dataset = tf.data.Dataset.from_tensor_slices((features, target))
dataset = dataset.batch(batch_size, drop_remainder=True)

#Checking sample data size
example_input_batch, example_target_batch = next(iter(dataset))
print("This is sample input_batch shape and target_batch shape: ",example_input_batch.shape, example_target_batch.shape)

flight = Flight()
sample_output = flight(example_input_batch)

for l in flight.layers:
    print(l.name, l.trainable)

sample_output = flight(example_input_batch)

optimizer = tf.keras.optimizers.Adam()
loss_calc = tf.keras.losses.MeanAbsoluteError()

# Saving checkpoints
checkpoint_dir = './training_checkpoints'
checkpoint_prefix = os.path.join(checkpoint_dir, "ckpt")
checkpoint = tf.train.Checkpoint(optimizer=optimizer,
                                flight = flight)

@tf.function
def train_step(inp, targ):
    loss = 0

    with tf.GradientTape() as tape:
        predicted = flight(inp)

        loss = loss_calc(targ, predicted)
        
        variables = flight.variables

        gradients = tape.gradient(loss, variables)

        optimizer.apply_gradients(zip(gradients, variables))

    return loss

epochs = 5
for epoch in range(epochs):
        start = time.time()

#         flight_hidden = flight.initialize_hidden_state()
        total_loss = 0

        for (batch, (inp, targ)) in enumerate(dataset.take(steps_per_epoch)):
            batch_loss = train_step(inp, targ)
            total_loss += batch_loss

        # saving (checkpoint) the model every 2 epochs
        checkpoint.save(file_prefix = checkpoint_prefix)

        print('Epoch {} Loss {:.4f}'.format(epoch + 1,
                                        total_loss / steps_per_epoch))
        print('Time taken for 1 epoch {} sec\n'.format(time.time() - start))

```

I've added a sample dataset for reproducing the problem. 
[tensorflow_issue.csv.zip](https://github.com/tensorflow/tensorflow/files/4223717/tensorflow_issue.csv.zip)

**Comment by pandrey-fr** [comment]
Your model's call method returns the output of `layer4(x)`, meaning that the first three layers have no effect on the predictions, hence the absence of gradients.

You probablye wanted to write :
```python
def call(self, x):
    output = self.linear1(x)
    output = self.linear2(output)
    output = self.linear3(output)
    output = self.linear4(output)
    return output
```
so that all four layers are being used sequentially in processing the inputs.
